### PR TITLE
refactor: simplified ascendance, moved oracle away

### PR DESCRIPTION
**Pull request by kupermind**
-  Simplified ascendance, moved oracle away.

### CODE REVIEW

File: contracts/BuyBackBurner.sol [hunk 1]
```diff
@@ -1,6 +1,9 @@
 // SPDX-License-Identifier: MIT
 pragma solidity ^0.8.28;
 
+import {IUniswapV3} from "./interfaces/IUniswapV3.sol";
+import {TickMath} from "./libraries/TickMath.sol";
```
Comment (kupermind) [inline contracts/BuyBackBurner.sol:5]: Moving oracle part here as discussed. No code change.

File: contracts/MemeArbitrum.sol [hunk 2]
```diff
@@ -19,10 +19,9 @@ contract MemeArbitrum is MemeFactory {
     ) MemeFactory(_olas, _nativeToken, _uniV3PositionManager, _buyBackBurner, _minNativeTokenValue) {}
 
     /// @dev Allows diverting first x collected funds to a launch campaign.
-    /// @param amount Amount of native token to convert to OLAS and burn.
-    /// @return adjustedAmount Adjusted amount of native token to convert to OLAS and burn.
-    function _launchCampaign(uint256 amount) internal override pure returns (uint256 adjustedAmount) {
-        return amount;
+    /// @return Adjusted amount of native token to convert to OLAS and burn.
+    function _launchCampaign() internal override view returns (uint256) {
+        return scheduledForAscendance;
```
Comment (kupermind) [inline contracts/MemeArbitrum.sol:24]: Tis really needs to return just `scheduledForAscendance`

File: contracts/MemeBase.sol [hunk 3]
```diff
@@ -70,25 +69,30 @@ contract MemeBase is MemeFactory {
...
-    function _launchCampaign(uint256 amount) internal override returns (uint256 adjustedAmount) {
+    function _launchCampaign() internal override returns (uint256 adjustedAmount) {
+        require(scheduledForAscendance >= LIQUIDITY_AGNT, "Not enough to cover launch campaign");
+
+        // Unleash the campaign token
         _MAGA();
-        adjustedAmount = amount - LIQUIDITY_AGNT;
+
+        // scheduledForAscendance might also increase during the pool creation
+        adjustedAmount = scheduledForAscendance - LIQUIDITY_AGNT;
```
Comment (kupermind) [inline contracts/MemeBase.sol:94]: `amount` is not needed anymore, read directly from the state.

File: contracts/MemeFactory.sol [hunk 4]
```diff
@@ -185,7 +185,7 @@ abstract contract MemeFactory {
         uint256 priceX96 = (amount1 * 1e18) / amount0;
 
         // Calculate the square root of the price ratio in X96 format
-        uint160 sqrtPriceX96 = uint160((FixedPointMathLib.sqrt(priceX96) * 2**96) / 1e9);
+        uint160 sqrtPriceX96 = uint160((FixedPointMathLib.sqrt(priceX96) * (1 << 96)) / 1e9);
```
Comment (kupermind) [inline contracts/MemeFactory.sol:188]: Same result without power calculation

File: contracts/MemeFactory.sol [hunk 5]
```diff
@@ -219,59 +219,14 @@ abstract contract MemeFactory {
         }
     }
 
-    function _getTwapFromOracle(address pool) internal view returns (uint256 priceX96) {
```
Comment (kupermind) [inline contracts/MemeFactory.sol:222]: Moved out to the `BuyBackBurner`

File: contracts/MemeFactory.sol [hunk 6]
```diff
@@ -582,16 +536,8 @@ abstract contract MemeFactory {
         require(_locked == 1, "Reentrancy guard");
         _locked = 2;
 
-        uint256 amount = scheduledForAscendance;
+        uint256 amount = _launched > 0 ? scheduledForAscendance : _launchCampaign();
```
Comment (kupermind) [inline contracts/MemeFactory.sol:539]: This is a major update and is much cleaner. Either return `scheduledForAscendance` or launch campaign based on `scheduledForAscendance` as well.

**Review by DavidMinarsch — APPROVED** [review]

**Pull request merged**